src/main.cpp

```diff
@@ -575,7 +574,178 @@ void affect_nearby_friends_with_alarm(struct Thing *traptng)
+      {
+        thing->health -= game.boulder_reduce_health_wall;
+      }
+      slide_thing_against_wall_at(thing, &pos, blocked_flags);
+      if ( blocked_flags & SlbBloF_WalledX )
+      {
+        angle = thing->move_angle_xy;
+        if ( (angle) && ( (angle <= 0x200u) || (angle > 0x600u) ) )
+        {
+          unsigned short y = *(unsigned short *)&thing->mappos.y.stl;
+          *(unsigned short *)&pos2.x.stl = *(unsigned short *)&thing->mappos.x.stl;
+          *(unsigned short *)&pos2.z.stl = 0;
+          *(unsigned short *)&pos2.y.stl = y - 3 * speed;
+          *(unsigned short *)&pos2.z.stl = get_thing_height_at(thing, &pos2);
+          new_angle = (thing_in_wall_at(thing, &pos2) < 1) ? 0 : 0x400;
+        }
+        else
+        {
+          *(unsigned short *)&pos2.x.stl = *(unsigned short *)&thing->mappos.x.stl;
+          *(unsigned short *)&pos2.z.stl = 0;
+          *(unsigned short *)&pos2.y.stl = *(unsigned short *)&thing->mappos.y.stl + 3 * speed;
+          *(unsigned short *)&pos2.z.stl = get_thing_height_at(thing, &pos2);
+          new_angle = (thing_in_wall_at(thing, &pos2) < 1) ? 0x400 : 0;
+        }
+      }
+      else if ( blocked_flags & SlbBloF_WalledY )
+      {
+        angle = thing->move_angle_xy;
+        if ( (angle) && (angle <= 0x400u) ) 
+        {
+          *(unsigned short *)&pos2.z.stl = 0;
+          *(unsigned short *)&pos2.y.stl = *(unsigned short *)&thing->mappos.y.stl;
+          *(unsigned short *)&pos2.x.stl = *(unsigned short *)&thing->mappos.x.stl + 3 * speed;
+          *(unsigned short *)&pos2.z.stl = get_thing_height_at(thing, &pos2);
+          new_angle = (thing_in_wall_at(thing, &pos2) < 1) ? 512 : 1536;
+        }
+        else
+        {
+          unsigned short x = *(unsigned short *)&thing->mappos.x.stl;
+          *(unsigned short *)&pos2.z.stl = 0;
+          *(unsigned short *)&pos2.y.stl = *(unsigned short *)&thing->mappos.y.stl;
+          *(unsigned short *)&pos2.x.stl = x - 3 * speed;
+          *(unsigned short *)&pos2.z.stl = get_thing_height_at(thing, &pos2);
+          new_angle = (thing_in_wall_at(thing, &pos2) < 1) ? 1536 : 512;
+        }
+      }
+      else
+      {
+        ERRORLOG("Cannot find boulder wall hug angle!");
+        new_angle = 0;
+      }
+      thing->move_angle_xy = new_angle;
+    }
+  }
+  angle = thing->move_angle_xy;
+  *(unsigned short *)&thing->velocity.x.stl = shotst->speed
+                                       * lbSinTable[angle] >> 16;
+  *(unsigned short *)&thing->velocity.y.stl = -(shotst->speed * lbCosTable[angle] >> 8) >> 8;
+  return 0;
+}
+
+long process_boulder_collision(struct Thing *thing, struct Coord3d *pos, int a3, int a4)
+// Function at 493690. Original name unknown.
+{ 
+  unsigned short i = (thing->clipbox_size_xy >> 1);
+  unsigned long pos_x = (pos->x.val + i * a3) >> 8;
+  unsigned long pos_y = (i * a4 + pos->y.val) >> 8;
+  MapSlabCoord slb_x = *((unsigned long *)&map_to_slab + pos_x);
```

Comment:
Would this be the same as `subtile_slab_fast(pos->x.stl.num);`?

Reply:
I don't know; I don't really understand how the original code works. I just know that it does.

Reply:
It is.